From 6a6a96e3ee657e71260c6ab8f545e9710ffddb7d Mon Sep 17 00:00:00 2001
From: Andy Duan <fugang.duan@nxp.com>
Date: Wed, 5 Apr 2017 15:30:59 +0800
Subject: [PATCH 1632/5242] MLK-14618 net: fec: fixed-link don't need phy
 fixup setting

commit  d792389b91be9123efbf3276e8052c87cb71bfc9 from
https://source.codeaurora.org/external/imx/linux-imx.git

In MAC-MAC fixed link case, it don't need phy fixup setting.

Signed-off-by: Fugang Duan <fugang.duan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/freescale/fec.h      |    1 +
 drivers/net/ethernet/freescale/fec_main.c |   17 ++++++++++++-----
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/freescale/fec.h b/drivers/net/ethernet/freescale/fec.h
index 601b33b..f983a2d 100644
--- a/drivers/net/ethernet/freescale/fec.h
+++ b/drivers/net/ethernet/freescale/fec.h
@@ -558,6 +558,7 @@ struct fec_enet_private {
 	phy_interface_t	phy_interface;
 	struct device_node *phy_node;
 	int	link;
+	bool	fixed_link;
 	int	full_duplex;
 	int	speed;
 	struct	completion mdio_done;
diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 7b190c3..8b41c16 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -2110,11 +2110,15 @@ static int fec_enet_mii_init(struct platform_device *pdev)
 	fep->mii_bus->parent = &pdev->dev;
 
 	node = of_get_child_by_name(pdev->dev.of_node, "mdio");
-	err = of_mdiobus_register(fep->mii_bus, node);
-	if (node)
+	if (node) {
+		err = of_mdiobus_register(fep->mii_bus, node);
 		of_node_put(node);
-	else if (fep->phy_node)
+	} else if (fep->phy_node && !fep->fixed_link) {
 		err = -EPROBE_DEFER;
+	} else {
+		err = mdiobus_register(fep->mii_bus);
+	}
+
 	if (err)
 		goto err_out_free_mdiobus;
 
@@ -3588,6 +3592,7 @@ static void fec_enet_of_parse_stop_mode(struct platform_device *pdev)
 			goto failed_phy;
 		}
 		phy_node = of_node_get(np);
+		fep->fixed_link = true;
 	}
 	fep->phy_node = phy_node;
 
@@ -3729,8 +3734,10 @@ static void fec_enet_of_parse_stop_mode(struct platform_device *pdev)
 	if (ret)
 		goto failed_register;
 
-	fep->fixups = of_fec_enet_parse_fixup(np);
-	fec_enet_register_fixup(ndev);
+	if (!fep->fixed_link) {
+		fep->fixups = of_fec_enet_parse_fixup(np);
+		fec_enet_register_fixup(ndev);
+	}
 
 	device_init_wakeup(&ndev->dev, fep->wol_flag &
 			   FEC_WOL_HAS_MAGIC_PACKET);
-- 
1.7.9.5

